helm: add imagePullSecrets for helm charts

fungaren opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

Currently the Helm chart does not contain a `imagePullSecrets` option when you are using private container registry, this is very inconvenient. This PR add this option for both CephFS and RBD.

Provide some context for the reviewer

For example, csi-driver-nfs already support this:

https://github.com/kubernetes-csi/csi-driver-nfs/blob/862cb036e3f0d7c3fe28d2b127a6f96335b15309/charts/v4.3.0/csi-driver-nfs/values.yaml#L127-L131

https://github.com/kubernetes-csi/csi-driver-nfs/blob/862cb036e3f0d7c3fe28d2b127a6f96335b15309/charts/v4.3.0/csi-driver-nfs/templates/csi-nfs-controller.yaml#L20-L23

## Is there anything that requires special attention ##

To test this feature, you can use `helm template` and `kubectl apply --dry-run` to examine the YAML result and the syntax. For example:

```text
~/ceph-csi/charts$ helm template -n default ceph-csi-rbd ./ceph-csi-cephfs \
--set provisioner.imagePullSecrets[0].name=aaaaa \
--set nodeplugin.imagePullSecrets[0].name=bbbbb | \
kubectl apply -f - --validate='strict' --dry-run=client

serviceaccount/ceph-csi-rbd-ceph-csi-cephfs-nodeplugin created (dry run)
serviceaccount/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
configmap/ceph-config created (dry run)
configmap/ceph-csi-config created (dry run)
clusterrole.rbac.authorization.k8s.io/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
clusterrolebinding.rbac.authorization.k8s.io/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
role.rbac.authorization.k8s.io/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
rolebinding.rbac.authorization.k8s.io/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
service/ceph-csi-rbd-ceph-csi-cephfs-nodeplugin-http-metrics created (dry run)
service/ceph-csi-rbd-ceph-csi-cephfs-provisioner-http-metrics created (dry run)
daemonset.apps/ceph-csi-rbd-ceph-csi-cephfs-nodeplugin created (dry run)
deployment.apps/ceph-csi-rbd-ceph-csi-cephfs-provisioner created (dry run)
csidriver.storage.k8s.io/cephfs.csi.ceph.com configured (dry run)
```

Do you have any questions?

Is the change backward compatible?

Are there concerns around backward compatibility?

Provide any external context for the change, if any.

For example:

* Kubernetes links that explain why the change is required
* CSI spec related changes/catch-up that necessitates this patch
* golang related practices that necessitates this change

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Fixes: #issue_number

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic commented:
Thanks!

Could you please add support for NFS as well? Also force-push your updated/rebased commit and include your `signed-off-by` approval. The subject of the commit should also start with `deploy: ` or `helm: `.

fungaren commented:
@nixpanic Well, but it seems that it does not provide a helm chart for NFS? Theres only `ceph-csi-cephfs` and `ceph-csi-rbd` in the `charts` directory.

nixpanic commented:
> @nixpanic Well, but it seems that it does not provide a helm chart for NFS? Theres only `ceph-csi-cephfs` and `ceph-csi-rbd` in the `charts` directory.

Ah, indeed, that seems to be missing :disappointed:

nixpanic commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -37,6 +37,10 @@ spec:
       dnsPolicy: ClusterFirstWithHostNet
+{{- if .Values.nodeplugin.imagePullSecrets }}
+      imagePullSecrets:
+{{ toYaml .Values.nodeplugin.imagePullSecrets | indent 8 -}}
```
Could you show in an example what value you would pass? I would expect something like this instead:

```yaml
 - name: {{ .Values.nodeplugin.imagePullSecrets }}
```

fungaren replied:
Here you should provide an array like:

```yaml
imagePullSecrets:
  - name: my-secret-a
  - name: my-secret-b
```

This part in fact is copied from csi-driver-nfs:

https://github.com/kubernetes-csi/csi-driver-nfs/blob/862cb036e3f0d7c3fe28d2b127a6f96335b15309/charts/v4.3.0/csi-driver-nfs/values.yaml#L127-L131

https://github.com/kubernetes-csi/csi-driver-nfs/blob/862cb036e3f0d7c3fe28d2b127a6f96335b15309/charts/v4.3.0/csi-driver-nfs/templates/csi-nfs-controller.yaml#L20-L23

And It is a recommended way rather than: 

```
imagePullSecrets:
  - name: {{ .Values.nodeplugin.imagePullSecret }}
```

In this way you can not provide multiple secrets. And maybe there are other fields (no just `name` only).

nixpanic replied:
ok, thanks for the clarification!

Madhu-1 commented:
@fungaren can you please fix linter errors https://github.com/ceph/ceph-csi/actions/runs/5267388978/jobs/9526493600?pr=3906 and https://github.com/ceph/ceph-csi/pull/3906/checks?check_run_id=14256026729. you can follow https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#code-contribution-workflow to sign the commit.

nixpanic commented:
> @fungaren can you please fix linter errors https://github.com/ceph/ceph-csi/actions/runs/5267388978/jobs/9526493600?pr=3906 and https://github.com/ceph/ceph-csi/pull/3906/checks?check_run_id=14256026729. you can follow https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#code-contribution-workflow to sign the commit.

I think the failure is because there is no empty line between the subject of the commit, and the Signed-off-by line.

fungaren committed:
helm: add imagePullSecrets option

Currently the Helm chart does not contain a
imagePullSecrets option when you are using
private container registry, this is very inconvenient.
This PR add this option for both CephFS and RBD.

Signed-off-by: Garen Fang <fungaren@qq.com>

nixpanic commented:
@megifyio queue

nixpanic commented:
@mergifyio queue

nixpanic commented:
Hi @fungaren, could you login (with your GitHub account) on the Mergify Dashboard? It seems that it is required for some reason before Mergify wants to run tests on this PR.

https://dashboard.mergify.com/github/ceph/repo/ceph-csi/queues?branch=devel&pull=3906&queues=default

Once done, leave a comment `@mergifyio refresh` in this PR. Thanks!

fungaren commented:
Emm.... It sounds a bit strange... Anyway, I will try. 

But I don't have permission to access it.

![screenshot-20230615-194802](https://github.com/ceph/ceph-csi/assets/8206464/c8446510-f3c3-4b01-9efc-23d6d48cdc0a)

Rakshith-R commented:
@fungaren Can you just rebase on top of devel now `git pull --rebase upstream devel` and force push the pr? 
that should work too AFAIK.

Rakshith-R commented:
@mergifyio queue

Rakshith-R commented:
@mergifyio refresh

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio requeue

nixpanic commented:
Thanks @fungaren , I think Mergify is finally also ready to accept+merge this PR.
